zerver/actions/message_send.py

```diff
@@ -858,11 +861,58 @@ def get_active_presence_idle_user_ids(
     return filter_presence_idle_user_ids(user_ids)
 
 
-@transaction.atomic(savepoint=False)
+def raw_insert_message_with_idempotency_key(
+    send_request: SendMessageRequest,
+    idempotency_key: str,
+) -> None:
+    """
+    Insert a single Message in db, using Idempotency-Key header.
+    """
+
+    # Get Message db column names.
+    # These fields are filtered in the same way django bulk_create does it.
```

Comment:
Relevant lines in djago implementation are [this](https://github.com/django/django/blob/main/django/db/models/query.py#L802) and [this ](https://github.com/django/django/blob/main/django/db/models/query.py#L824)

Reply:
Tip for the future: Press `y` on the keyboard to make the link into a permalink -- otherwise, the line number is embedded into the link, but the code in `main` will continue to move the lines around, making the link not go where you intend when someone looks at it later/